### ISSUE TITLE
False positive UnusedVariable

### ISSUE THREAD

**Issue by ArtemGoutsoul** [vimeo/psalm#4435]
False positive UnusedVariable in psalm 4.0.1 

```php
<?php

// false positive: INFO: UnusedParam - 3:32 - Param $year is never referenced in this method
function year_not_empty(string $year): bool
{
    $year_not_empty = (bool)$year;
    return $year_not_empty;
}

// false positive: INFO: UnusedVariable - 11:5 - Variable $number is never referenced
function coint_toss(): bool
{
    $number = rand(0, 1);
    $number_not_zero = (bool)$number;
    return $number_not_zero;
}

// no error
function year_not_empty2(string $year): bool
{
    return (bool)$year;
}

// no error
function coint_toss2(): bool
{
    $number = rand(0, 1);
    return (bool)$number;
}
```

Link: https://psalm.dev/r/6aaf133f48

Also the suggested fix breaks things e.g.:
```
function coint_toss(): bool
{
    rand(0, 1);
    $number_not_zero = (bool)$number;
    return $number_not_zero;
}
```

One more slightly more realistic case:
https://psalm.dev/r/d35da68f5a
```
function apply_operator($subject, string $operator, $value)
{
	$subject = (float)$subject;
	$value = (float)str_replace(',', '.', (string)$value);
	switch ($operator) {
		case '+':
			$subject += $value;
			break;
		case '-':
			$subject -= $value;
			break;
		default:
			if ($operator) {
				throw new Exception("Operator '$operator' is not supported!");
			}
			throw new Exception("Operator is required");
	}
	return $subject;
}
```

```
Psalm output (using commit fee5eab): 

INFO: UnusedParam - 9:25 - Param $subject is never referenced in this method
```

**Comment by muglug** [comment]
Right, `(float), `(array)`, `(object)` and `(bool)` casts were not flow-sensitive. Now they are.
